src/index/src/fulltext_index/create/tantivy.rs

```
@@ -0,0 +1,316 @@
+        let text_field = schema_builder.add_text_field(TEXT_FIELD_NAME, TEXT);
+        let schema = schema_builder.build();
+
+        let mut index = Index::create_in_dir(path, schema).context(TantivySnafu)?;
+        index.settings_mut().docstore_compression = Compressor::Zstd(ZstdCompressor::default());
+        index.set_tokenizers(Self::build_tokenizer(&config));
+
+        let memory_limit = Self::sanitize_memory_limit(memory_limit);
+
+        let writer = SingleSegmentIndexWriter::new(index, memory_limit).context(TantivySnafu)?;
+        Ok(Self {
+            writer: Some(writer),
+            text_field,
+        })
+    }
+
+    fn build_tokenizer(config: &Config) -> TokenizerManager {
+        let mut builder = match config.analyzer {
+            Analyzer::English => TextAnalyzer::builder(SimpleTokenizer::default()).dynamic(),
+            Analyzer::Chinese => TextAnalyzer::builder(JiebaTokenizer {}).dynamic(),
+        };
+
+        if !config.case_sensitive {
+            builder = builder.filter_dynamic(LowerCaser);
+        }
+
+        let tokenizer = builder.build();
+        let tokenizer_manager = TokenizerManager::new();
+        tokenizer_manager.register("default", tokenizer);
+        tokenizer_manager
+    }
+
+    fn sanitize_memory_limit(memory_limit: usize) -> usize {
+        // Port from tantivy::indexer::index_writer::{MEMORY_BUDGET_NUM_BYTES_MIN, MEMORY_BUDGET_NUM_BYTES_MAX}
+        const MARGIN_IN_BYTES: usize = 1_000_000;
+        const MEMORY_BUDGET_NUM_BYTES_MIN: usize = ((MARGIN_IN_BYTES as u32) * 15u32) as usize;
+        const MEMORY_BUDGET_NUM_BYTES_MAX: usize = u32::MAX as usize - MARGIN_IN_BYTES;
+
+        memory_limit.clamp(MEMORY_BUDGET_NUM_BYTES_MIN, MEMORY_BUDGET_NUM_BYTES_MAX)
+    }
+}
+
+#[async_trait]
+impl FulltextIndexCreator for TantivyFulltextIndexCreator {
+    async fn push_text(&mut self, text: &str) -> Result<()> {
+        if let Some(writer) = &mut self.writer {
```

Maybe return an error when writer is already finished?